pkg/syncer/types/commontypes.go

```diff
@@ -58,31 +60,38 @@ func InitConfigInfo(ctx context.Context) (*ConfigInfo, error) {
 // GetVirtualCenterInstance returns the vcenter object singleton.
 // It is thread safe.
 func GetVirtualCenterInstance(ctx context.Context, configTypes *ConfigInfo) (*cnsvsphere.VirtualCenter, error) {
-	onceForVirtualCenter.Do(func() {
-		log := logger.GetLogger(ctx)
-		var vcconfig *cnsvsphere.VirtualCenterConfig
-		vcconfig, vcerror := cnsvsphere.GetVirtualCenterConfig(ctx, configTypes.Cfg)
-		if vcerror != nil {
-			log.Errorf("failed to get VirtualCenterConfig. Err: %+v", vcerror)
-			return
-		}
+	if atomic.LoadUint32(&vCenterInitialized) == 0 {
+		mutex.Lock()
+		defer mutex.Unlock()
+		if vCenterInitialized == 0 {
```

Comment:
I didn't find any load and store operations for boolean in atomic package.
https://golang.org/pkg/sync/atomic/

`uint32` can be used as a substitute.

Reply:
@chethanv28 `k8sorchestrator` also used `uint32`. I referred to `Once.Do` implementation for adopting this approach.

Reply:
> Why not use a boolean ? What is the advantage of using uint32 ?

+1, without the `atomic..` part we can just use a bool

Reply:
We need the atomic load and store to ensure we don't run into any race conditions.

`if boolVar == false` and `boolVar = true` - these operations are not truly atomic.
Now consider this - One go routine is setting the value of boolVar to true but right before this value is written to memory, second go routine reads a cache line and sees the value as false. In that case, both routines will end up creating an instance.

Reply:
The suggestion is to remove the `atomic.Load()` and instead `lock-verify-unlock`, where the `verify` step can be to compare a boolean variable rather than an integer. 

BTW the exact scenario you're describing exists with `atomic.Load()`. It's not a solution to any race condition because multiple threads can load `vCenterInitialized` with 0.

Reply:
> because multiple threads can load vCenterInitialized with 0.

That is also true. 
But I'm curious why `Once.Do` implementation is using atomic operations then, as @shalini-b mentioned?
https://golang.org/src/sync/once.go?s=1137:1164#L55

Found some related discussion here but it didn't conclude with any solid reasoning.
https://github.com/golang/go/issues/22104

Anyhow, I will remove the `atomic.Load()` and use a bool variable instead. But will keep the check before locking, as for most invocations, the instance would have already been initialized. So it will be `verify-lock-verify-unlock`.

Reply:
@chethanv28 @RaunakShah Changes done. PTAL.